src/asm/x86/transform/forward.rs

```diff
@@ -296,81 +296,89 @@ unsafe fn round_shift_array_avx2(arr: &mut [I32X8], size: usize, bit: i8) {
-          ))),
-          shift,
-        )
+#[allow(clippy::identity_op, clippy::erasing_op)]
+#[target_feature(enable = "avx2")]
+unsafe fn fwd_txfm2d_avx2<T: Coefficient>(
+  input: &[i16], output: &mut [T], stride: usize, tx_size: TxSize,
+  tx_type: TxType, bd: usize,
+) {
+  // Note when assigning txfm_size_col, we use the txfm_size from the
+  // row configuration and vice versa. This is intentionally done to
+  // accurately perform rectangular transforms. When the transform is
+  // rectangular, the number of columns will be the same as the
+  // txfm_size stored in the row cfg struct. It will make no difference
+  // for square transforms.
+  let txfm_size_col = tx_size.width();
+  let txfm_size_row = tx_size.height();
+
+  let col_class = SizeClass1D::from_length(txfm_size_col);
+  let row_class = SizeClass1D::from_length(txfm_size_row);
+
+  let mut tmp: AlignedArray<[I32X8; 64 * 64 / 8]> =
+    AlignedArray::uninitialized();
+  let buf = &mut tmp.array[..txfm_size_col * (txfm_size_row / 8).max(1)];
+  let tx_in = &mut [I32X8::zero(); 64];
+  let tx_out = &mut [I32X8::zero(); 64];
+  let cfg = Txfm2DFlipCfg::fwd(tx_type, tx_size, bd);
+
+  let txfm_func_col = get_func_i32x8(cfg.txfm_type_col);
+  let txfm_func_row = get_func_i32x8(cfg.txfm_type_row);
+
+  // Columns
+  for cg in (0..txfm_size_col).step_by(8) {
+    let shift = cfg.shift[0] as u8;
+    #[target_feature(enable = "avx2")]
+    #[inline]
+    unsafe fn load_columns(input_ptr: *const i16, shift: u8) -> I32X8 {
+      // TODO: load 64-bits for x4 wide columns
+      shift_left(
+        I32X8::new(_mm256_cvtepi16_epi32(_mm_loadu_si128(
+          input_ptr as *const _,
+        ))),
+        shift,
+      )
+    }
+    if cfg.ud_flip {
+      // flip upside down
+      for r in 0..txfm_size_row {
+        let input_ptr =
+          input[(txfm_size_row - r - 1) * stride + cg..].as_ptr();
+        tx_in[r] = load_columns(input_ptr, shift);
       }
-      if cfg.ud_flip {
-        // flip upside down
-        for r in 0..txfm_size_row {
-          let input_ptr =
-            input[(txfm_size_row - r - 1) * stride + cg..].as_ptr();
-          tx_in[r] = load_columns(input_ptr, shift);
-        }
-      } else {
-        for r in 0..txfm_size_row {
-          let input_ptr = input[r * stride + cg..].as_ptr();
-          tx_in[r] = load_columns(input_ptr, shift);
-        }
+    } else {
+      for r in 0..txfm_size_row {
+        let input_ptr = input[r * stride + cg..].as_ptr();
+        tx_in[r] = load_columns(input_ptr, shift);
       }
+    }
 
-      txfm_func_col(tx_in, tx_out);
+    txfm_func_col(tx_in, tx_out);
 
-      round_shift_array_avx2(tx_out, txfm_size_row, -cfg.shift[1]);
+    round_shift_array_avx2(tx_out, txfm_size_row, -cfg.shift[1]);
 
-      for rg in (0..txfm_size_row).step_by(8) {
-        if txfm_size_row >= 8 && txfm_size_col >= 8 {
+    // Transpose the array. Select the appropriate method to do so.
+    match (row_class, col_class) {
```

Comment:
You may move this further up, you may use hawktrace to compare before and after.

Reply:
tx_out is overwritten in each loop iteration.

Reply:
I wanted move the transposes out and set the right transpose outside the loop. I can do myself later :)